remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/RemoteQueryEngine.java

```diff
@@ -89,14 +90,19 @@ protected Query makeTypeQuery(Query query, String targetEntityName) {
    }
 
    @Override
-   protected CacheQuery<?> makeCacheQuery(IckleParsingResult<Descriptor> ickleParsingResult, Query luceneQuery) {
-      CustomTypeMetadata customTypeMetadata = () -> {
-         IndexingMetadata indexingMetadata = ickleParsingResult.getTargetEntityMetadata().getProcessedAnnotation(IndexingMetadata.INDEXED_ANNOTATION);
-         return indexingMetadata != null ? indexingMetadata.getSortableFields() : Collections.emptySet();
-      };
-      IndexedTypeMap<CustomTypeMetadata> queryMetadata = IndexedTypeMaps.singletonMapping(ProtobufValueWrapper.INDEXING_TYPE, customTypeMetadata);
-      HSQuery hSearchQuery = getSearchFactory().createHSQuery(luceneQuery, queryMetadata);
-      return ((SearchManagerImpl) getSearchManager()).getQuery(hSearchQuery);
+   protected CacheQuery<?> makeCacheQuery(IckleParsingResult<Descriptor> ickleParsingResult, Query luceneQuery, IndexedQueryMode queryMode, Map<String, Object> namedParameters) {
+      IndexingMetadata indexingMetadata = ickleParsingResult.getTargetEntityMetadata().getProcessedAnnotation(IndexingMetadata.INDEXED_ANNOTATION);
+      final Set<String> sortableFields = indexingMetadata != null ? indexingMetadata.getSortableFields() : Collections.emptySet();
+      IndexedTypeMap<CustomTypeMetadata> queryMetadata = IndexedTypeMaps.singletonMapping(ProtobufValueWrapper.INDEXING_TYPE, () -> sortableFields);
+      RemoteQueryDefinition queryDefinition;
+      if (queryMode == IndexedQueryMode.BROADCAST) {
+         queryDefinition = new RemoteQueryDefinition(ickleParsingResult.getQueryString());
+      } else {
+         queryDefinition = new RemoteQueryDefinition(getSearchFactory().createHSQuery(luceneQuery, queryMetadata));
+      }
+      queryDefinition.setNamedParameters(namedParameters);
+      SearchManagerImplementor searchManager = getSearchManager().unwrap(SearchManagerImplementor.class);
```

Comment:
I see ```getSearchManager().unwrap(SearchManagerImplementor.class)``` being called a few times. I think at this point it is better to modify QueryEngine.getSearchManager() to return SearchManagerImplementor and do the unwrapping only first time it is obtained.

Reply:
ok